### PR TITLE
KAFKA-15093: Add 3.4 and 3.5 Streams upgrade system tests

### PR DESCRIPTION
**Pull request by mimaison**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mimaison** [comment]
I copied what was done in previous releases. I'm unsure if any APIs have changed or what to really look for.
@mjsax @vvcephei Can you take a look?

File: streams/upgrade-system-tests-34/src/test/java/org/apache/kafka/streams/tests/SmokeTestUtil.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    static ProcessorSupplier<Object, Object, Void, Void> printProcessorSupplier(final String topic, final String name) {
+        return () -> new ContextualProcessor<Object, Object, Void, Void>() {
+            private int numRecordsProcessed = 0;
+            private long smallestOffset = Long.MAX_VALUE;
+            private long largestOffset = Long.MIN_VALUE;
+
+            @Override
+            public void init(final ProcessorContext<Void, Void> context) {
+                super.init(context);
+                System.out.println("[3.3] initializing processor: topic=" + topic + " taskId=" + context.taskId());
```
Comment (mjsax) [inline streams/upgrade-system-tests-34/src/test/java/org/apache/kafka/streams/tests/SmokeTestUtil.java]:
```suggestion
                System.out.println("[3.4] initializing processor: topic=" + topic + " taskId=" + context.taskId());
```

File: streams/upgrade-system-tests-34/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    @SuppressWarnings("unchecked")
+    public static void main(final String[] args) throws Exception {
+        if (args.length < 1) {
+            System.err.println("StreamsUpgradeTest requires one argument (properties-file) but provided none");
+        }
+        final String propFileName = args[0];
+
+        final Properties streamsProperties = Utils.loadProps(propFileName);
+
+        System.out.println("StreamsTest instance started (StreamsUpgradeTest v3.3)");
```
Comment (mjsax) [inline streams/upgrade-system-tests-34/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java]:
```suggestion
        System.out.println("StreamsTest instance started (StreamsUpgradeTest v3.4)");
```

File: streams/upgrade-system-tests-34/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        kStream.to("fk-result", Produced.with(stringSerde, stringSerde));
+    }
+
+    private static <KIn, VIn, KOut, VOut> ProcessorSupplier<KIn, VIn, KOut, VOut> printProcessorSupplier(final String topic) {
+        return () -> new ContextualProcessor<KIn, VIn, KOut, VOut>() {
+            private int numRecordsProcessed = 0;
+
+            @Override
+            public void init(final ProcessorContext<KOut, VOut> context) {
+                System.out.println("[3.3] initializing processor: topic=" + topic + "taskId=" + context.taskId());
```
Comment (mjsax) [inline streams/upgrade-system-tests-34/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java]:
```suggestion
                System.out.println("[3.4] initializing processor: topic=" + topic + "taskId=" + context.taskId());
```

File: streams/upgrade-system-tests-35/src/test/java/org/apache/kafka/streams/tests/SmokeTestUtil.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    static ProcessorSupplier<Object, Object, Void, Void> printProcessorSupplier(final String topic, final String name) {
+        return () -> new ContextualProcessor<Object, Object, Void, Void>() {
+            private int numRecordsProcessed = 0;
+            private long smallestOffset = Long.MAX_VALUE;
+            private long largestOffset = Long.MIN_VALUE;
+
+            @Override
+            public void init(final ProcessorContext<Void, Void> context) {
+                super.init(context);
+                System.out.println("[3.3] initializing processor: topic=" + topic + " taskId=" + context.taskId());
```
Comment (mjsax) [inline streams/upgrade-system-tests-35/src/test/java/org/apache/kafka/streams/tests/SmokeTestUtil.java]:
```suggestion
                System.out.println("[3.5] initializing processor: topic=" + topic + " taskId=" + context.taskId());
```

File: streams/upgrade-system-tests-35/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    @SuppressWarnings("unchecked")
+    public static void main(final String[] args) throws Exception {
+        if (args.length < 1) {
+            System.err.println("StreamsUpgradeTest requires one argument (properties-file) but provided none");
+        }
+        final String propFileName = args[0];
+
+        final Properties streamsProperties = Utils.loadProps(propFileName);
+
+        System.out.println("StreamsTest instance started (StreamsUpgradeTest v3.3)");
```
Comment (mjsax) [inline streams/upgrade-system-tests-35/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java]:
```suggestion
        System.out.println("StreamsTest instance started (StreamsUpgradeTest v3.5)");
```

File: streams/upgrade-system-tests-35/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        kStream.to("fk-result", Produced.with(stringSerde, stringSerde));
+    }
+
+    private static <KIn, VIn, KOut, VOut> ProcessorSupplier<KIn, VIn, KOut, VOut> printProcessorSupplier(final String topic) {
+        return () -> new ContextualProcessor<KIn, VIn, KOut, VOut>() {
+            private int numRecordsProcessed = 0;
+
+            @Override
+            public void init(final ProcessorContext<KOut, VOut> context) {
+                System.out.println("[3.3] initializing processor: topic=" + topic + "taskId=" + context.taskId());
```
Comment (mjsax) [inline streams/upgrade-system-tests-35/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java]:
```suggestion
                System.out.println("[3.5] initializing processor: topic=" + topic + "taskId=" + context.taskId());
```

**Review by mjsax — COMMENTED** [review]
Thanks. It's really just a c&p (view minor updates required).

We should trigger a system test run before we merge to see it its working as expected.

**Comment by mimaison** [comment]
@mjsax I also noticed https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignorConfiguration.java#L180
This was not updated for 3.4, does this need updating too?

**Comment by vvcephei** [comment]
Good catch, @mimaison .

It looks like we could keep adding "upgrade from" versions on each release, but unless there's a compatibility issue, we don't need to. If you look at `git blame`, you can see that, in the past, we didn't add those cases one at a time, but in chunks when we had a compatibility problem to manage.

**Comment by mjsax** [comment]
Thanks for confirming @vvcephei -- as a matter of fact, we released KP-904 with 3.5, so it seems to be an issue. Let me do a PR to address this and also add a test to get a guard in place.

**Comment by mimaison** [comment]
Are there compatibility issues with 3.4 or 3.5?

**Comment by mjsax** [comment]
With KIP-904 we changes some serialization format inside an internal repartition topic (cf https://cwiki.apache.org/confluence/display/KAFKA/KIP-904%3A+Kafka+Streams+-+Guarantee+subtractor+is+called+before+adder+if+key+has+not+changed)

To upgrade KS correctly, you need to do two rolling bounces (first one with `upgrade.from = "3.4"` -- or whatever version you are using), and second one with the config set back to `null`.

Thus, it seems, while the upgrade path for KP-904 itself works, the `AssignorConfiguration` you mentioned would not recognize `"3.4"` and thus crash Kafka Streams.

We did not get to finish the system test for this upgrade (it's a WIP PR: https://github.com/apache/kafka/pull/13656), that I assume would have caught this issue. Guess I should have insistent to get the system test in place before the release (that is totally my fault)

So basically, if one uses `KTable.groupBy().aggregate()` and want to upgrade from 3.4 to 3.5, they cannot do it without resetting the application.

\cc @fqaiser94 @cadonna

**Comment by mjsax** [comment]
Did a PR: https://github.com/apache/kafka/pull/14103

**Comment by mjsax** [comment]
I just uploaded the artifact for 3.5.1 to the S3 bucket we use for system tests -- can we add 3.5.1 in this PR right away (or even wait for 3.5.2 -- we might need a hotfix for this upgrade issue...?)

**Comment by divijvaidya** [comment]
> can we add 3.5.1 in this PR right away

I have added system tests to use 3.5.1 in https://github.com/apache/kafka/pull/14069

**Comment by mimaison** [comment]
System tests look better with https://github.com/apache/kafka/pull/14103. In addition to a few flaky tests, I got 3 failures:
1. 
    ```
    Module: kafkatest.tests.streams.streams_broker_bounce_test
    Class:  StreamsBrokerBounceTest
    Method: test_broker_type_bounce
    Arguments:
    {
      "broker_type": "controller",
      "failure_mode": "hard_shutdown",
      "metadata_quorum": "ZK",
      "num_threads": 1,
      "sleep_time_secs": 120
    }
    ```
2. 
    ```
    Module: kafkatest.tests.streams.streams_broker_bounce_test
    Class:  StreamsBrokerBounceTest
    Method: test_broker_type_bounce
    Arguments:
    {
      "broker_type": "leader",
      "failure_mode": "hard_shutdown",
      "metadata_quorum": "ZK",
      "num_threads": 1,
      "sleep_time_secs": 120
    }
    ```
3.
    ```
    Module: kafkatest.tests.streams.streams_upgrade_test
    Class:  StreamsUpgradeTest
    Method: test_rolling_upgrade_with_2_bounces
    Arguments:
    {
      "from_version": "3.5.0",
      "to_version": "3.6.0-SNAPSHOT"
    }
    ```
I still need to look at the logs of all failures but at first sight 1. and 2. don't seem directly related.

**Review by mjsax — APPROVED** [review]
Thanks!

**Comment by mimaison** [comment]
This depends on https://github.com/apache/kafka/pull/14103 so we need to get that merged first

**Comment by mjsax** [comment]
@mimaison Merged the hotfix PR.

**Comment by mimaison** [comment]
I rebased on top of trunk and reran the system tests in our CI. Many of the Streams upgrade system tests are flaky but I've seen all of them pass at least once apart from `kafkatest.tests.streams.streams_upgrade_test.test_rolling_upgrade_with_2_bounces_from_version=3_5_1_to_version=3_6_0-SNAPSHOT`. This one consistently failed across ~10 runs.

From my limited Streams understanding I don't see any obvious issues in the logs. @mjsax can you take a look and see if it's just really flaky or if there's a real issue. Is is passing in your CI?  I've uploaded the logs from the last run to my Apache home directory: https://home.apache.org/~mimaison/streams-upgrade-failure.zip

**Comment by mjsax** [comment]
We also observed failing tests -- it's on our TODO list but might take some time until we get to it :( -- I'll keep you posted.

**Comment by mimaison** [comment]
@mjsax These missing tests are starting to pile up. We're already missing them for 3.4 and 3.5, and 3.6 will be out shortly. If we were to break compatibility in Streams how would we notice it if we don't have these tests? Actually do we know whether the failures are caused by the tests or by issues in Streams? Figuring this out seems relatively high priority for me.

**Comment by mjsax** [comment]
I identified one fundamental problem in 3.3 release -- a bug was introduced that fundamentally breaks KS systems tests using that version. This PR (https://github.com/apache/kafka/commit/fa0324485b1e76388bb15ea98170cfb9202aab1f) changes `ProcessorParameter` and allows `this.processorSupplier` to be `null` what breaks `toString()` method on the same class.

It was fixed via https://github.com/apache/kafka/commit/0de037423bef416abd654bd8b90b642d56eb33ea but it's only included in 3.4.1 release -- while it was backported to `3.3` branch, too, we never did a bug fix release afterwards and thus the system test that pull release binaries only cannot pickup the cherry-pick.

**Comment by mjsax** [comment]
Another issue just introduced in `trunk` -> https://github.com/apache/kafka/pull/14490

**Comment by mimaison** [comment]
Thanks @mjsax for following up. So what's the path forward? Are we able to merge this or do we need to release new bugfix versions for 3.3 and 3.4 (and 3.5?)?

**Comment by mjsax** [comment]
I just opened a PR to fix `trunk` -- if this PR is merged, I can cherry-pick it to older branches, and you can rebase this PR so we can verify everything is in order and we can merge this one.

https://github.com/apache/kafka/pull/14539

**Comment by mjsax** [comment]
>  while it was backported to 3.3 branch, too, we never did a bug fix release afterwards and thus the system test that pull release binaries only cannot pickup the cherry-pick

I could not reproduce the issue any longer and did more digging -- it think I hit this issue previously because we did not bump the `Dockerfile` to use the latest bug fix versions. Looking into the fix for this, it's actually shipped with 3.3.2 release, so I think we are fine.

**Comment by mimaison** [comment]
I rebased this PR on trunk and rerun the Streams system tests and still got a bunch of failures:
```
kafkatest.tests.streams.streams_upgrade_test.test_rolling_upgrade_with_2_bounces_from_version=0_10_0_1_to_version=3_7_0-SNAPSHOT
kafkatest.tests.streams.streams_cooperative_rebalance_upgrade_test.test_upgrade_to_cooperative_rebalance_upgrade_from_version=0_10_0_1
kafkatest.tests.streams.streams_cooperative_rebalance_upgrade_test.test_upgrade_to_cooperative_rebalance_upgrade_from_version=0_10_1_1
kafkatest.tests.streams.streams_cooperative_rebalance_upgrade_test.test_upgrade_to_cooperative_rebalance_upgrade_from_version=0_10_2_2
kafkatest.tests.streams.streams_cooperative_rebalance_upgrade_test.test_upgrade_to_cooperative_rebalance_upgrade_from_version=0_11_0_3
kafkatest.tests.streams.streams_cooperative_rebalance_upgrade_test.test_upgrade_to_cooperative_rebalance_upgrade_from_version=1_0_2
kafkatest.tests.streams.streams_cooperative_rebalance_upgrade_test.test_upgrade_to_cooperative_rebalance_upgrade_from_version=1_1_1
kafkatest.tests.streams.streams_cooperative_rebalance_upgrade_test.test_upgrade_to_cooperative_rebalance_upgrade_from_version=2_0_1
kafkatest.tests.streams.streams_cooperative_rebalance_upgrade_test.test_upgrade_to_cooperative_rebalance_upgrade_from_version=2_1_1
kafkatest.tests.streams.streams_cooperative_rebalance_upgrade_test.test_upgrade_to_cooperative_rebalance_upgrade_from_version=2_2_2
kafkatest.tests.streams.streams_cooperative_rebalance_upgrade_test.test_upgrade_to_cooperative_rebalance_upgrade_from_version=2_3_1
kafkatest.tests.streams.streams_broker_down_resilience_test.test_streams_runs_with_broker_down_initially_metadata_quorum=ISOLATED_KRAFT
kafkatest.tests.streams.streams_upgrade_test.test_rolling_upgrade_with_2_bounces_from_version=0_10_1_1_to_version=3_7_0-SNAPSHOT
kafkatest.tests.streams.streams_upgrade_test.test_rolling_upgrade_with_2_bounces_from_version=0_11_0_3_to_version=3_7_0-SNAPSHOT
kafkatest.tests.streams.streams_upgrade_test.test_rolling_upgrade_with_2_bounces_from_version=0_10_2_2_to_version=3_7_0-SNAPSHOT
kafkatest.tests.streams.streams_upgrade_test.test_rolling_upgrade_with_2_bounces_from_version=1_0_2_to_version=3_7_0-SNAPSHOT
kafkatest.tests.streams.streams_upgrade_test.test_rolling_upgrade_with_2_bounces_from_version=1_1_1_to_version=3_7_0-SNAPSHOT
kafkatest.tests.streams.streams_upgrade_test.test_rolling_upgrade_with_2_bounces_from_version=3_5_1_to_version=3_7_0-SNAPSHOT
kafkatest.tests.streams.streams_broker_compatibility_test.test_compatible_brokers_eos_alpha_enabled_broker_version=0_11_0_3
kafkatest.tests.streams.streams_broker_compatibility_test.test_compatible_brokers_eos_disabled_broker_version=0_11_0_3
kafkatest.tests.streams.streams_broker_compatibility_test.test_fail_fast_on_incompatible_brokers_broker_version=0_10_1_1
kafkatest.tests.streams.streams_broker_compatibility_test.test_fail_fast_on_incompatible_brokers_broker_version=0_10_2_2
kafkatest.tests.streams.streams_broker_compatibility_test.test_fail_fast_on_incompatible_brokers_if_eos_v2_enabled_broker_version=0_11_0_3
```
Tests can be a bit flaky in our CI but many of these seem to fail consistently. @mjsax did you get a clean run or do you see these tests failing in your CI as well?

**Comment by mjsax** [comment]
> I rebased this PR on trunk 

https://github.com/apache/kafka/pull/14539 is not merged yet... So that's expected I guess. -- We actually also just found another bug that was fixed in the meantime: https://github.com/apache/kafka/pull/14555

Let's see if we can push 14539 over the finish line first... Will ping here again after that's done; sorry that it take so long, but we are on it. Promised.

**Comment by mjsax** [comment]
Summary about your test results:
 - `streams_upgrade_test` expected to still fail because 14539 was not merged yet
 - `streams_cooperative_rebalance_upgrade_test` -- seems to be flaky -- after re-running them multiple times, I got each one to pass at least once locally (we should still try to stabilize this test, but I don't think it's a broker to move forward with this PR)
 - `streams_broker_down_resilience_test` -- this test was broken by a recent commit (https://github.com/apache/kafka/commit/fcac880fd54efbec3fe385000cf990a19972dafa) which changes a log message the test expects -- I'll add a fix to 14539 for it
 - `streams_broker_compatibility_test` -- all test passed locally on the first run; not sure -- if we indeed see issues on Jenkins we need to figure out why it's flaky

**Comment by mjsax** [comment]
Updated https://github.com/apache/kafka/pull/14539 and re-triggered a system test run. Let's hope for the best :)

**Comment by mjsax** [comment]
Merged the other PR. Btw: should we split this PR into two, one to add 3.4 and one to add 3.5? So it's easier to cherry-pick? (Adding 3.5 should only go into `3.6` branch, but adding 3.4 should got into `3.6` and `3.5` branches.)

**Comment by mimaison** [comment]
@mjsax Good call, I've split this PR in 2:
- 3.4: https://github.com/apache/kafka/pull/14601
- 3.5: https://github.com/apache/kafka/pull/14602

Closing this PR
